Add set_fact_persistent action and module.

alikins opened the pull request:
##### SUMMARY

Used just like set_fact, except facts set with set_fact_persistent
will be stored in the fact cache if fact caching is enabled.

set_fact only sets facts in the non_persistent_fact_cache, so they
are lost between invocations.

The implementations here are pretty much exact copies of set_fact.
The module likely needs to do that, but the action plugin could in theory
import set_fact and subclass it (all it needs is a different action name...)

<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
modules/utilities/logic/set_fact.py
plugins/strategy/
plugins/action/

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (set_fact_persistent c1e8b05ade) last updated 2017/08/02 10:00:27 (GMT -400)
  config file = /home/adrian/src/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]


```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

alikins committed:
Add 'cacheable' param to  set_fact action and module.

Used just like set_fact, except facts set with cacheable: true
will be stored in the fact cache if fact caching is enabled.

set_fact normally only sets facts in the non_persistent_fact_cache, so they
are lost between invocations.

bcoca commented:
why not add a `cacheable` property to set_facts

alikins commented:
> why not add a cacheable property to set_facts

Mainly because this approach doesn't require changes to plugins/strategy/*
(strategies could check for a property/attr on the action instead of the explicit check for 'action == "set_fact"' currently used in strategy)

bcoca commented:
using a 'cacheable' flag would be more flexible than the current 'set_fact' match, other modules could create non persistent/persistent facts

alikins commented:
@bcoca can add a cacheable Attribute to Task, and set it from set_fact.py (or normal.py) and check it in strategy

set_fact module assumes all params passed to it are key=value pairs however, so adding a 'cacheable' arg to the module could cause conflicts (seemingly even 'args').

If we avoid the potential name conflict and dont add a module arg for skip_facts, then it will have to default to false and we need another module that could take a 'cacheable' arg. But at the moment, all other modules are already cacheable, but would still need a set_fact_persistent that enables cacheable and set_fact.

Or we could make 'cacheable' a prohibited fact name to use with set_fact...

bcoca commented:
anyone setting 'args' as a var is already creating issues for themselves, other 'raw' parameter modules allow options (see creates= in shell/command) so it would not be a first.

my worry about introducing a new module is also that it introduces a new precedence step, while introducing a toggle between the 2 types of facts does not.

alikins commented:
Adding the 'cacheable' arg and a Task.cacheable works, but I'm not sure the precedence is doing the right thing. In particular, for a case where a play does a non-persistent set_fact, then a persistent set_fact
after that.

``` yaml
---
- name: the first play
  hosts: localhost
  tasks:
    - name: show foobar fact before
      debug:
        var: ansible_foobar

    - name: set a ansible_foobar fact with out cacheable
      set_fact:
        ansible_foobar: 'blippy-no-cacheable'

    - name: debug ansible_foobar no-cachable
      debug:
        var: ansible_foobar

    - name: set a ansible_foobar fact with cachable
      set_fact:
          ansible_foobar: 'blippy-cacheable-1'
          cacheable: true

    - name: debug ansible_foobar cacheable 1
      debug:
        var: ansible_foobar
```
Resuls:
```
[newswoop:F25:ansible (set_fact_persistent %)]$ ANSIBLE_CONFIG=ansible-set-fact-persistent.cfg 
PLAYBOOK: set_fact_2.yml **********************************************************************************
1 plays in set_fact_2.yml

PLAY [the first play] *************************************************************
META: ran handlers

TASK [show foobar fact before] ********************************************************************************
task path: /home/adrian/src/ansible/set_fact_2.yml:5
ok: [localhost] => {
    "ansible_foobar": "blippy-cacheable-1", 
    "failed": false
}

TASK [set a ansible_foobar fact with out cacheable] ********************************************************************************
task path: /home/adrian/src/ansible/set_fact_2.yml:9
ok: [localhost] => {
    "ansible_facts": {
        "ansible_foobar": "blippy-no-cacheable"
    }, 
    "changed": false, 
    "failed": false
}

TASK [debug ansible_foobar no-cachable] ********************************************************************************
task path: /home/adrian/src/ansible/set_fact_2.yml:13
ok: [localhost] => {
    "ansible_foobar": "blippy-no-cacheable", 
    "failed": false
}

TASK [set a ansible_foobar fact with cachable] **********************************************************************************
task path: /home/adrian/src/ansible/set_fact_2.yml:17
ok: [localhost] => {
    "ansible_facts": {
        "ansible_foobar": "blippy-cacheable-1", 
        "cacheable": true
    }, 
    "changed": false, 
    "failed": false
}

TASK [debug ansible_foobar cacheable 1] ********************************************************************************
task path: /home/adrian/src/ansible/set_fact_2.yml:22
ok: [localhost] => {
    "ansible_foobar": "blippy-no-cacheable", 
    "failed": false
}

```

The non_persistent_fact_cache is a higher precedence which explains why/how it happens, but it seems kind of odd and unexpected.

alikins committed:
update set_facts docs

alikins commented:
Updated. 

Not sure if the behavior on regular set_fact is correct. Logic was updated so that set_fact always sets non_persistent_facts and if cacheable, also sets host_facts. 

That seems to do the right thing, but not sure about the details of the precedence. 

Without the change, the behavior when using caching and non-caching set_fact on the same fact name is a little odd. A fact set non caching, and then later set with caching, the non caching value wins because of the higher precedence of non_persistent facts. Adding to non persistent facts and host_facts seems to DWIM there, but unsure of consequences elsewhere.

bcoca commented:
its odd, but it is documented, set_fact wins over 'actual facts' .. the real difference being cacheable/non-cacheable.

I don't think this works well as task keyword as most tasks don't return facts, only module that return 'ansible_facts' do.

alikins committed:
use 'ansible_facts_cacheable' in module/actions result

alikins committed:
pop fact cacheable related items out of args/results

We dont want to use 'ansible_facts_cacheable' result item
or 'cacheable' arg as actual facts, so pop them out of the
dicts.

alikins commented:
> using a 'cacheable' flag would be more flexible than the current 'set_fact' match, other modules could create non persistent/persistent facts

> I don't think this works well as task keyword as most tasks don't return facts, only module that return 'ansible_facts' do.

So, just a '_ansible_fact_cacheable' in module/action return and checked by strategy?

jmighion commented:
I like this approach. It would allow me to cache facts that are important to me, but probably not important enough to make upstream changes to a specific fact module and eliminate maintaining local changes to modules over the course of Ansible releases.

isopy commented:
Definitely gotta +1 a cacheable option for set_facts.

bpleines commented:
+1

harrytruman commented:
+1

abadger commented on lib/ansible/plugins/strategy/__init__.py:
```diff
@@ -511,12 +511,14 @@ def parent_handler_match(target_handler, handler_name):
                                 for target_host in host_list:
                                     self._variable_manager.set_host_variable(target_host, var_name, var_value)
                         else:
+                            cacheable = result_item.pop('ansible_facts_cacheable', True)
```
I don't mess around in strategy often so I don't know if this is normal but,, why is the default here True but False everywhere else?

bcoca replied:
because `set_fact` is only module in which this was not true, all other 'ansible_facts' should be cacheable by default.
